File: app/src/main/java/com/qianlei/jiaowu/ui/view/TermChooseView.java
package com.qianlei.jiaowu.ui.view;

import android.content.Context;
import android.util.AttributeSet;
import android.view.LayoutInflater;
import android.widget.AdapterView;
import android.widget.LinearLayout;
import android.widget.Spinner;

import androidx.annotation.Nullable;

import com.qianlei.jiaowu.R;
import com.qianlei.jiaowu.utils.DateUtil;

/**
 * @author qianlei
 */
public class TermChooseView extends LinearLayout {
    private Spinner yearSpinner;
    private Spinner termSpinner;

    public TermChooseView(Context context) {
        this(context, null);
    }

    public TermChooseView(Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
        LayoutInflater.from(context).inflate(R.layout.term_choose_view, this);
        yearSpinner = findViewById(R.id.spinner_year);
        termSpinner = findViewById(R.id.spinner_term);
        //设置选中项为当前学期
        yearSpinner.setSelection(DateUtil.getCurYear() - 2017);
        termSpinner.setSelection(DateUtil.getCurTerm() - 1);
    }

    public void setItemSelectedListener(AdapterView.OnItemSelectedListener listener) {
        yearSpinner.setOnItemSelectedListener(listener);
        termSpinner.setOnItemSelectedListener(listener);
    }

    /**
     * 获取选择的学期
     *
     * @return 当前用户选择的学期
     */
    public String getTerm() {
        String[] terms = getResources().getStringArray(R.array.term);
        String term = terms[termSpinner.getSelectedItemPosition()];
        return formatTerm(term);
    }

    /**
     * 获取选择的学年
     *
     * @return 当前用户选择的学年
     */
    public String getYear() {
        String[] years = getResources().getStringArray(R.array.year);
        return years[yearSpinner.getSelectedItemPosition()].substring(0, 4);
    }

    /**
     * 格式化term
     * 传到教务系统的学期格式需要转换
     *
     * @param term 转换前的学期
     * @return 转换后的学期
     */
    private String formatTerm(String term) {
        switch (term) {
            case "2":
                return "12";
            case "3":
                return "16";
            case "1":
            default:
                return "3";
        }
    }
}

File: app/src/main/java/com/qianlei/jiaowu/ui/view/TermChooseView.kt
package com.qianlei.jiaowu.ui.view

import android.content.Context
import android.util.AttributeSet
import android.view.LayoutInflater
import android.widget.AdapterView.OnItemSelectedListener
import android.widget.LinearLayout
import android.widget.Spinner
import com.qianlei.jiaowu.R
import com.qianlei.jiaowu.entity.Term

/**
 * @author qianlei
 */
class TermChooseView @JvmOverloads constructor(context: Context?, attrs: AttributeSet? = null) : LinearLayout(context, attrs) {
    private val yearSpinner: Spinner
    private val termSpinner: Spinner
    fun setItemSelectedListener(listener: OnItemSelectedListener?) {
        yearSpinner.onItemSelectedListener = listener
        termSpinner.onItemSelectedListener = listener
    }

    /**
     * 获取选择的学期
     *
     * @return 当前用户选择的学期
     */
    val term: Term
        get() {
            val years = resources.getStringArray(R.array.year)
            val terms = resources.getStringArray(R.array.term)
            val term = terms[termSpinner.selectedItemPosition]
            val year = years[yearSpinner.selectedItemPosition].substring(0, 4)
            return Term(year, formatTerm(term))
        }

    /**
     * 格式化term
     * 传到教务系统的学期格式需要转换
     *
     * @param term 转换前的学期
     * @return 转换后的学期
     */
    private fun formatTerm(term: String): String {
        return when (term) {
            "2" -> "12"
            "3" -> "16"
            "1" -> "3"
            else -> "3"
        }
    }

    init {
        LayoutInflater.from(context).inflate(R.layout.term_choose_view, this)
        yearSpinner = findViewById(R.id.spinner_year)
        termSpinner = findViewById(R.id.spinner_term)
        val years = resources.getStringArray(R.array.year)
        //设置选中项为当前学期
        yearSpinner.setSelection(Term.getNowTerm().year.toInt() - years[0].substring(0, 4).toInt())
        termSpinner.setSelection(Term.getNowTerm().term.toInt() - 1)
    }
}